release-23.2: server/license: Limit enterprise trial license usage to once per cluster

spilchen opened the pull request:
Backport 1/1 commits from #131422.

/cc @cockroachdb/release

---

Previously, clusters could install a second enterprise trial license. To prevent misuse, we are now restricting the use of this license type to only once per cluster. This update adds the necessary tracking mechanism.

A new key has been introduced in the KV system key range to record the number of enterprise trial licenses used, ensuring the data persists across restarts.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-40416
Release note (general change): Attempting to install a second enterprise trial license on the same cluster will now fail.
Release justification: This work is part of the CockroachDB core deprecation.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/131973)
<!-- Reviewable:end -->

spilchen committed:
server/license: Limit enterprise trial license usage to once per cluster

Previously, clusters could install a second enterprise trial license. To
prevent misuse, we are now restricting the use of this license type to only
once per cluster. This update adds the necessary tracking mechanism.

A new key has been introduced in the KV system key range to record the number
of enterprise trial licenses used, ensuring the data persists across restarts.

This change will be backported to versions 24.2, 24.1, 23.2, and 23.1.

Epic: CRDB-39988
Closes: CRDB-40416
Release note (general change): Attempting to install a second enterprise trial
license on the same cluster will now fail.

rafiss reviewed (approved):
> Reviewed 10 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/131973)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @spilchen)



<!-- Sent from Reviewable.io -->
